File: GASEngineSource/Engine/SkeletonManager.js
//Skeleton Manager
GASEngine.SkeletonManager = function ()
{
    this.boneList = [];
    this.skeletonEntity = [];
    this.updatePairs = [];
    this.scale = 1.0;
    this.min_size = Number.MAX_SAFE_INTEGER;
    this.max_size = 0;
    this.skeletonHelper = null;

    this.unitSphereMesh = new GASEngine.SphereMesh(1.0);
    this.unitSphereMesh.submitToWebGL();

    this.unitOctahedronMesh = new GASEngine.OctahedronMesh(new GASEngine.Vector3(0,0,1), 0.5);
    this.unitOctahedronMesh.submitToWebGL();

    GASEngine.SkeletonManager.Instance = this;
};

GASEngine.SkeletonManager.prototype =
{
    constructor: GASEngine.SkeletonManager,

    init: function()
    {
        if (this.boneList.length > 0)
        {
            this.createSkeleton();

            this.skeletonHelper = GASEngine.EntityFactory.Instance.create();
            this.skeletonHelper.type = 'helper';
            this.skeletonHelper.name = 'skeletonHelper';
            this.skeletonHelper.uniqueID = GASEngine.generateUUID();

            this.skeletonHelper.children = this.skeletonEntity;
            this.skeletonHelper.enable = true;
            this.skeletonEntity.forEach(v=>v.parent = this.skeletonHelper);
            //console.log(skeletonHelper);
            //var bone = this.createBone(this.bonelist[0]);
            return this.skeletonHelper;
        }
        else
        {
            console.error("This model has no skeletons!");  
            return null;  
        }
    },

    finl: function()
    {
        if (this.skeletonHelper !== null) {
            if (this.skeletonHelper.parent !== null) {
                this.skeletonHelper.parent.removeChild(this.skeletonHelper);
            }
            GASEngine.EntityFactory.Instance.destroy(this.skeletonHelper);
            this.skeletonHelper = null;
        }
        // skeletonEntity均为helper的child，上面的destroy会清理掉的。
        this.boneList.length = 0;
        this.skeletonEntity.length = 0;
        this.updatePairs.length = 0;
        this.scale = 1.0;
    },

    update: function()
    {
        if (this.min_size > this.max_size) {
            this.updateMinMax();
        }
        if (this.min_size <= this.max_size) {
            this.updatePairs.forEach(v =>{
                if (v[0] == 1) {
                    this.applyTransformToBone(v[1], v[2]);
                } else if (v[0] == 2) {
                    this.applyTransformToJoint(v[1], v[2]);
                }
            });
        }
    },

    appendBones: function (bones)
    {
        for (var i = 0; i < bones.length; ++i)
        {
            if (!this.boneList.includes(bones[i]))
            {
                this.boneList.push(bones[i]);
            }
        }
    },

    createJoint: function (entity)
    {
        var jointEntity = GASEngine.EntityFactory.Instance.create();
        jointEntity.name = entity.name;
        jointEntity.type = 'helper';
        jointEntity.uniqueID = GASEngine.generateUUID();
        jointEntity.boneID = entity.uniqueID;

        var meshFilterComponent = GASEngine.ComponentFactory.Instance.create('meshFilter');
        meshFilterComponent.uniqueID = GASEngine.generateUUID();
        meshFilterComponent.setMesh(this.unitSphereMesh);

        var meshRenderComponent = GASEngine.ComponentFactory.Instance.create('meshRenderer');
        var material = GASEngine.MaterialFactory.Instance.create('lambert');
        //material.culling = GASEngine.Material.CullingOff;
        meshRenderComponent.addMaterial(material);

        jointEntity.addComponent(meshFilterComponent);
        jointEntity.addComponent(meshRenderComponent);

        return jointEntity;
    },

    createBone: function (child, parent)
    {
        //var skeleton = this.bonelist[i].parent.clone();
        var boneEntity = GASEngine.EntityFactory.Instance.create();
        boneEntity.name = child.name;
        boneEntity.type = 'helper';
        boneEntity.setLocalTranslation(new GASEngine.Vector3(0,0,0));
        //skeleton.setLocalQuaternion(this.bonelist[i].getLocalQuaternion());

        //skeleton.parent = this.root;
        boneEntity.uniqueID = GASEngine.generateUUID();
        //skeleton.boneID = this.bonelist[i].uniqueID;

        var meshFilterComponent = GASEngine.ComponentFactory.Instance.create('meshFilter');
        meshFilterComponent.uniqueID = GASEngine.generateUUID();
        var dir = child.getLocalTranslation().clone();//.sub(this.bonelist[i].parent.getWorldTranslation().clone());
        meshFilterComponent.setMesh(this.unitOctahedronMesh);

        var meshRenderComponent = GASEngine.ComponentFactory.Instance.create('meshRenderer');
        var material = GASEngine.MaterialFactory.Instance.create('lambert');
        //material.culling = GASEngine.Material.CullingOff;
        meshRenderComponent.addMaterial(material);

        boneEntity.addComponent(meshFilterComponent);
        boneEntity.addComponent(meshRenderComponent);

        return boneEntity;
    },

    createSkeleton: function ()
    {
        for (var i = 0; i < this.boneList.length; ++i)
        {
            var src_entity = this.boneList[i];
            if (src_entity.parent && this.boneList.includes(src_entity.parent))
            {
                var boneEntity = this.createBone(src_entity, src_entity.parent);
                this.skeletonEntity.push(boneEntity);
                this.updatePairs.push([1, src_entity, boneEntity]);
                var jointEntity = this.createJoint(src_entity);
                this.skeletonEntity.push(jointEntity);
                this.updatePairs.push([2, src_entity, jointEntity]);
            }
        }
    },

    updateMinMax: function ()
    {
        this.min_size = Number.MAX_SAFE_INTEGER;
        this.max_size = 0;
        for (var i = 0; i < this.boneList.length; ++i)
        {
            if (this.boneList[i].parent && this.boneList.includes(this.boneList[i].parent))
            {
                // var distance = this.boneList[i].getLocalTranslation().length();
                var distance_vector = this.boneList[i].getWorldTranslation().clone();
                distance_vector.sub(this.boneList[i].parent.getWorldTranslation());
                var distance_world = distance_vector.length();
                // console.log('distance is ' + distance);
                if (distance_world > 0) {
                    this.min_size = Math.min(this.min_size, distance_world);
                    this.max_size = Math.max(this.max_size, distance_world);
                }
            }
        }
    },

    applyTransformToBone: function(src_entity, dest_entity) {
        var pos = src_entity.getWorldTranslation().clone();
        pos.sub(src_entity.parent.getWorldTranslation());
        var distance = pos.length();
        var xy_scale = this.scale * Math.clamp(distance, this.min_size, this.max_size) / 15.0;
        var model_scale = new GASEngine.Matrix4();
        model_scale.scale(new GASEngine.Vector3(xy_scale, xy_scale, distance));
        pos.normalize();
        var model_rotate = new GASEngine.Matrix4();
        if (pos.z < 0.9999) {
            var rotate_axis = new GASEngine.Vector3(-pos.y, pos.x, 0).normalize();
            var cos_theta_d2 = Math.sqrt((1 + pos.z) / 2);
            var sin_theta_d2 = Math.sqrt(1 - cos_theta_d2 * cos_theta_d2);
            var quat = new GASEngine.Quaternion(sin_theta_d2 * rotate_axis.x,
                 sin_theta_d2 * rotate_axis.y,
                 sin_theta_d2 * rotate_axis.z,
                 cos_theta_d2);
            model_rotate.makeRotationFromQuaternion(quat);
        }
        var final_matrix = new GASEngine.Matrix4();
        final_matrix.setPosition(src_entity.parent.getWorldTranslation());
        final_matrix.multiplyMatrices(final_matrix, model_rotate);
        final_matrix.multiplyMatrices(final_matrix, model_scale);
        dest_entity.setLocalMatrix(final_matrix);
        dest_entity.matrixWorld.copy(final_matrix);
    },

    applyTransformToJoint: function(src_entity, dest_entity) {
        var pos = src_entity.getWorldTranslation().clone();
        pos.sub(src_entity.parent.getWorldTranslation());
        var distance = pos.length();

        var radius = this.scale * Math.clamp(distance, this.min_size, this.max_size) / 2 / 15.0;
        var model_scale = new GASEngine.Matrix4();
        model_scale.scale(new GASEngine.Vector3(radius, radius, radius));
        var final_matrix = new GASEngine.Matrix4();
        final_matrix.setPosition(src_entity.parent.getWorldTranslation());
        final_matrix.multiplyMatrices(final_matrix, model_scale);
        dest_entity.setLocalMatrix(final_matrix);
        dest_entity.matrixWorld.copy(final_matrix);
    }
};

File: GASEngineSource_OpenGL/src/ecs/enhancement/skeleton_manager.cpp
#include "skeleton_manager.h"
#include "data_types/geometry.h"
#include "data_types/material_factory.h"
#include "ecs/component/mesh_filter_component.h"
#include "ecs/component/mesh_renderer_component.h"

pEntity SkeletonManager::GetEntity() const { return manager_entity; }

void SkeletonManager::Update() {
    if (min_scale > max_scale) {
        UpdateMinMax();
    }
    if (max_scale >= min_scale) {
        for (const auto& m : update_system) m();
    }
}

void SkeletonManager::UpdateMinMax() {
    // distance range is [min_scaling, max_scaling]
    min_scale = std::numeric_limits<float>::max();
    max_scale = 0;
    for (const auto& m : bone_set) {
        if (m->parent != nullptr && bone_set.count(m->parent) != 0) {
            float current_distance =
                glm::l2Norm(m->GetWorldTranslation() - m->parent->GetWorldTranslation());
            if (current_distance > 0.0f) {
                min_scale = std::min(current_distance, min_scale);
                max_scale = std::max(current_distance, max_scale);
            }
        }
    }
}

void SkeletonManager::CreateHelperEntity(const pEntity& parent) {
    if (manager_entity != nullptr) {
        manager_entity->parent->RemoveChild(manager_entity);
    }
    manager_entity = EntityFactory::Instance().CreateEntity(parent);
    manager_entity->enable = true;
    manager_entity->name = "skeleton_helper";
    manager_entity->is_helper = true;
}

void SkeletonManager::AppendBones(const std::vector<pEntity>& bone_list) {
    for (const auto& m : bone_list) {
        if (bone_set.count(m) == 0) {
            bone_set.emplace(m);
        }
    }
}

void SkeletonManager::AppendBone(const pEntity& bone_entity) {
    if (bone_set.count(bone_entity) == 0) {
        bone_set.emplace(bone_entity);
    }
}

void SkeletonManager::ApplyTransformToBone(const pEntity& src, const pEntity& dest) {
    assert(src->parent != nullptr);
    glm::vec3 vec = src->GetWorldTranslation() - src->parent->GetWorldTranslation();
    float distance = glm::l2Norm(vec);
    float xy_scale = scale * std::max(distance, min_scale) / 15.0f;
    glm::mat4 matrix_model_scale = glm::scale(glm::mat4(1.0f), {xy_scale, xy_scale, distance});

    glm::vec3 direction = glm::normalize(vec);
    glm::mat4 matrix_model_rotate(1.0f);
    if (std::abs(direction.z) < 0.99999) {
        glm::vec3 rotate_axis = glm::normalize(glm::vec3(-direction.y, direction.x, 0));
        float cos_theta_d_2 = std::sqrt((1 + direction.z) / 2);
        float sin_theta_d_2 = std::sqrt(1 - cos_theta_d_2 * cos_theta_d_2);
        glm::quat quat(cos_theta_d_2, sin_theta_d_2 * rotate_axis);
        Rotation r;
        r.FromQuat(quat);
        matrix_model_rotate = r.AsMat();
        // check should equals to pos;
        // glm::vec3 check = matrix_model_rotate * glm::vec4{0, 0, distance, 1};
        // auto diff = glm::abs(check - pos);
        // auto diff_val = diff.x + diff.y + diff.z;
    } else if (direction.z < 0) {
        matrix_model_rotate[0][0] = -1.0f;
        matrix_model_rotate[2][2] = -1.0f;
    }

    glm::mat4 matrix_world = glm::translate(glm::mat4(1.0f), src->parent->GetWorldTranslation());
    dest->SetLocalMatrix(matrix_world * matrix_model_rotate * matrix_model_scale);
    dest->UpdateWorldMatrix();
}

void SkeletonManager::ApplyTransformToJoint(const pEntity& src, const pEntity& dest) {
    glm::vec3 vec = src->GetWorldTranslation();
    if (src->parent != nullptr) {
        vec -= src->parent->GetWorldTranslation();
    }
    float distance = glm::l2Norm(vec);

    float radius = scale * std::clamp(distance, min_scale, (max_scale + min_scale) / 2.0f) / 12.0f;
    glm::mat4 matrix_model = glm::scale(glm::mat4(1.0f), {radius, radius, radius});
    glm::mat4 matrix_world = glm::translate(glm::mat4(1.0f), src->GetWorldTranslation());
    dest->SetLocalMatrix(matrix_world * matrix_model);
    dest->UpdateWorldMatrix();
}

void SkeletonManager::CreateSkeletonEntities(const pEntity& parent) {
    if (manager_entity == nullptr) {
        CreateHelperEntity(parent);
    }
    for (const auto& m : children) {
        manager_entity->RemoveChild(m);
    }
    children.clear();
    update_system.clear();
    for (const auto& m : bone_set) {
        if (m->parent != nullptr && bone_set.count(m->parent) != 0) {
            pEntity e_bone = CreateBone(m, m->parent);
            children.push_back(e_bone);
            update_system.push_back([m, e_bone, this]() { this->ApplyTransformToBone(m, e_bone); });
        }
        pEntity e_joint = CreateJoint(m);
        children.push_back(e_joint);
        update_system.push_back([m, e_joint, this]() { this->ApplyTransformToJoint(m, e_joint); });
    }
    for (const auto& m : children) {
        manager_entity->AddChild(m);
    }
}

pEntity SkeletonManager::CreateJoint(const pEntity& target) {
    pEntity joint = EntityFactory::Instance().CreateEntity(manager_entity);
    joint->name = "JOINT_" + target->name;
    joint->is_helper = true;
    joint->SetParent(target->parent);

    pComponent c;
    c = ComponentFactory::Instance().Create(kMeshFilter);
    auto mfc = std::dynamic_pointer_cast<MeshFilterComponent>(c);
    // auto sphere = SphereMesh::Generate(radius * scale);
    // sphere->SubmitToOpenGL();
    mfc->SetMesh(SphereMesh::UnitMesh());
    mfc->UpdateBBox();
    joint->AddComponent(mfc);

    c = ComponentFactory::Instance().Create(kMeshRenderer);
    auto mrc = std::dynamic_pointer_cast<MeshRendererComponent>(c);
    auto material = MaterialFactory::Instance().Create(kLambertianMaterial);
    material->LoadAllContent();
    material->SetDepthTest(false);
    mrc->AddMaterial(material);
    joint->AddComponent(mrc);

    return joint;
}

pEntity SkeletonManager::CreateBone(const pEntity& target, const pEntity& parent) {
    pEntity bone = EntityFactory::Instance().CreateEntity(manager_entity);
    bone->name = "BONE_" + target->name;
    bone->is_helper = true;
    bone->SetParent(parent);

    pComponent c;
    c = ComponentFactory::Instance().Create(kMeshFilter);
    auto mfc = std::dynamic_pointer_cast<MeshFilterComponent>(c);
    // auto sphere = PyramidMesh::Generate(target->GetLocalTranslation(), scale);
    // sphere->SubmitToOpenGL();
    mfc->SetMesh(PyramidMesh::UnitMesh());
    mfc->UpdateBBox();
    bone->AddComponent(mfc);

    c = ComponentFactory::Instance().Create(kMeshRenderer);
    auto mrc = std::dynamic_pointer_cast<MeshRendererComponent>(c);
    auto material = MaterialFactory::Instance().Create(kLambertianMaterial);
    material->LoadAllContent();
    material->SetDepthTest(false);
    mrc->AddMaterial(material);
    bone->AddComponent(mrc);

    return bone;
}
